### PR TITLE
Update "dirty" file only when `GitIsDirty` changes

### PR DESCRIPTION
**Pull request by AmoreCadenza**
This change fixes what looks like a bug in the current implementation of the `_GitRoot` target: namely, that the target always appends a `0` or `1`, on a new line, to the `_GitIsDirtyFile` every time said target runs, causing all other targets that depend on it as an input to always rerun as well. Ultimately, this breaks incremental builds of projects that rely on the `GitInfo` package.

With this change, the `_GitIsDirtyFile` behaves as its logic appears to have been intended to: instead of always appending the current value of `$(GitIsDirty)` to the file, the current value of `($GitIsDirty)` will replace the file's contents if and only if the value differs from what's already in the file. Otherwise, the file remains untouched and all other targets that use it as an input will function as expected in incremental builds.

This fixes or at least significantly mitigates the problems described in #183.

### CODE REVIEW

File: src/GitInfo/build/GitInfo.targets [hunk 1]
```diff
@@ -339,7 +339,7 @@
 			<GitIsDirty>0</GitIsDirty>
 		</PropertyGroup>
 
-		<WriteLinesToFile File="$(_GitIsDirtyFile)" Lines="$(GitIsDirty)" WriteOnlyWhenDifferent="true"/>
+		<WriteLinesToFile File="$(_GitIsDirtyFile)" Lines="$(GitIsDirty)" Overwrite="true" WriteOnlyWhenDifferent="true"/>
```
Comment (bddckr) [inline src/GitInfo/build/GitInfo.targets:342]: I believe this change makes this task append to the file once the file was created. `Overwrite` has to be `true` for `WriteOnlyWhenDifferent` to be used at all it seems:
https://github.com/dotnet/msbuild/blob/2e82c9f3f5feaa9444e67c4509c25f98fb1ea500/src/Tasks/FileIO/WriteLinesToFile.cs#L89-L95

Comment (AmoreCadenza) [reply src/GitInfo/build/GitInfo.targets:342]: @bddckr: I'm not sure I follow your comment. The current behavior, in my own investigation of this issue, is that this task appends a new line containing the value of `($GitIsDirty)` to the file every time the task is called.

This change, by adding `Overwrite="true"`, modifies this task's behavior so that the file's contents get replaced if and only if:

1. the file already exists and
2. the file's contents differ from the value of `$(GitIsDirty)`;

and, otherwise, the task no-ops.

What happens when the file does not exist  - create it and write the contents of `$(GitIsDirty)` into it - is, to the best of my knowledge, unaffected by this change.

Am I missing something? I can vouch for this change fixing incremental builds for the projects I integrated `GitInfo` into.

Comment (bddckr) [reply src/GitInfo/build/GitInfo.targets:342]: @AmoreCadenza Apologies, I must have read the diff as exactly the wrong way round, making me think you're breaking the functionality that you explained to be fixing 🤦

As we can see in the link to the source of the task that I posted earlier, your fix is basically what's turning on the `WriteOnlyWhenDifferent` behavior - because the source only ever even checks that property if `Overwrite` is also true.

tl;dr:
Before this change `WriteOnlyWhenDifferent` was never used as it's only ever used if `Overwrite` is turned on. With this change it is thus being used, ultimately giving the intended function of only writing _when needed_.

Consider this comment resolved! (Looks like I can't mark it as such, even though I'm the comment thread starter.)

Comment (AmoreCadenza) [reply src/GitInfo/build/GitInfo.targets:342]: Thanks for clearing that up, @bddckr! I'm glad we're on the same page. Let me know if there's anything else needed to get this merged. 😄

I see a button to resolve this conversation thread on my end so I'll try pressing it.

**Comment by Sergio0694** [comment]
Hey @devlooped, are there any concerns on this to get it merged? We also had to temporarily disable this package reference to reduce the build time, as this issue was breaking the incremental build and causing projects to be rebuilt unnecessarily. Would be great to have an updated version of the library with this fix, assuming the changes in here look right to you 😊
